src/mscorlib/src/System/Globalization/EncodingTable.Unix.cs

```diff
@@ -1,101 +1,162 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Text;
+
 namespace System.Globalization
 {
-    using System;
-    using System.Text;
-    using System.Collections;
-    using System.Collections.Generic;
-    using System.Runtime.CompilerServices;
-    using System.Runtime.InteropServices;
-    using System.Runtime.Versioning;
-    using System.Security;
-    using System.Threading;
-    using System.Diagnostics.Contracts;
-
     internal static class EncodingTable
     {
+        //
+        // This caches the data item which is indexed by the code page value.
+        //
+        private static Hashtable hashByCodePage = Hashtable.Synchronized(new Hashtable());
+
         // Return a list of all EncodingInfo objects describing all of our encodings
         [System.Security.SecuritySafeCritical]  // auto-generated
-        internal static unsafe EncodingInfo[] GetEncodings()
+        internal static EncodingInfo[] GetEncodings()
         {
-            // TODO: Implement this fully.
-            return new EncodingInfo[] {
-                new EncodingInfo(CodePageUtf7, "utf-7", "Unicode (UTF-7)"),
-                new EncodingInfo(CodePageUtf8, "utf-8", "Unicode (UTF-8)"),
-                new EncodingInfo(CodePageUtf16, "utf-16", "Unicode"),
-                new EncodingInfo(CodePageUtf16BE, "utf-16BE", "Unicode (Big-Endian)"),
-                new EncodingInfo(CodePageUtf32, "utf-32", "Unicode (UTF-32)"),
-            };
-        }
-    
-        /*=================================GetCodePageFromName==========================
-        **Action: Given a encoding name, return the correct code page number for this encoding.
-        **Returns: The code page for the encoding.
-        **Arguments:
-        **  name    the name of the encoding
-        **Exceptions:
-        **  ArgumentNullException if name is null.
-        **  internalGetCodePageFromName will throw ArgumentException if name is not a valid encoding name.
-        ============================================================================*/
-        
-        internal static int GetCodePageFromName(String name)
-        {
-            // TODO: Implement this fully.
-            switch (name)
-            {
-                case "utf-7":
-                    return CodePageUtf7; 
+            EncodingInfo[] arrayEncodingInfo = new EncodingInfo[encodingDataTableItems.Length];
 
-                case "utf-8":
-                    return CodePageUtf8;
+            for (int i = 0; i < encodingDataTableItems.Length; i++)
+            {
+                CodePageDataItem dataItem = encodingDataTableItems[i];
 
-                case "utf-16":
-                    return CodePageUtf16;
+                arrayEncodingInfo[i] = new EncodingInfo(dataItem.CodePage, dataItem.WebName,
```

Comment:
`EncodingInfo` is immutable. Should the instances be cached? (The array itself cannot be cached, because arrays can be changed, but the instances?). Or do callers of `Encoding.GetEncodings()` expect the instances to be unique references each time it is called? `EncodingInfo` overrides `Equals` and `GetHashCode()` (value semantics) so it seems unlikely the callers would care about unique references.